### PR TITLE
GSdx: Improve robustness of ComboBoxInit()

### PR DESCRIPTION
**Pull request by ssakash**
**Summary of changes**:

* Prevent comboboxes from reaching an indeterminate state **(Refer commit message for more details)**

### CODE REVIEW

File: plugins/GSdx/GSDialog.h [hunk 1]
```diff
@@ -51,7 +51,7 @@ class GSDialog
 	void SetText(UINT id, const char* str);
 	void SetTextAsInt(UINT id, int i);
 
-	void ComboBoxInit(UINT id, const vector<GSSetting>& settings, int32_t selectionValue, int32_t maxValue = INT32_MAX);
+	void ComboBoxInit(UINT id, const vector<GSSetting> settings, int32_t selectionValue, int32_t maxValue = INT32_MAX);
```
Comment (gregory38) [inline plugins/GSdx/GSDialog.h]: why did you remove the reference?

File: plugins/GSdx/GSDialog.cpp [hunk 2]
```diff
@@ -164,12 +164,27 @@ void GSDialog::SetTextAsInt(UINT id, int i)
...
+void GSDialog::ComboBoxInit(UINT id, const vector<GSSetting> settings, int32_t selectionValue, int32_t maxValue)
 {
+	if (settings.empty()) return;
+
 	HWND hWnd = GetDlgItem(m_hWnd, id);
+	bool currentsel_found = false;
 
 	SendMessage(hWnd, CB_RESETCONTENT, 0, 0);
 
+	for (const GSSetting i : settings)
```
Comment (gregory38) [inline plugins/GSdx/GSDialog.cpp]: better `const auto& i` or if you want the type `const GSSetting& i`

File: plugins/GSdx/GSDialog.h (hunk 1 above)
Comment (ssakash) [inline plugins/GSdx/GSDialog.h]: I probably overlooked the const reference I guess.

File: plugins/GSdx/GSDialog.cpp (hunk 2 above)
Comment (turtleli) [inline plugins/GSdx/GSDialog.cpp]: `std::find_if` could be used instead of a for loop.

Comment (ssakash) [inline plugins/GSdx/GSDialog.cpp]: > std::find_if could be used instead of a for loop.

Unless I'm wrong, I'd also have to define an unary function for the third parameter of ``find_if()`` in that case, is that fine?

Comment (gregory38) [inline plugins/GSdx/GSDialog.cpp]: Here some examples (because I played with the texture cache, I don't know if VS2013 support it).

```
plugins/GSdx/GSDevice.cpp-	auto compatible_texture = [&type, &format, &size, &msaa](const GSTexture* t) {
plugins/GSdx/GSDevice.cpp-		return (t->GetType() == type && t->GetFormat() == format && t->GetSize() == size && t->IsMSAA() == msaa);
plugins/GSdx/GSDevice.cpp-	};
plugins/GSdx/GSDevice.cpp-
plugins/GSdx/GSDevice.cpp:	const auto hit = std::find_if(m_pool.begin(), m_pool.end(), compatible_texture);
plugins/GSdx/GSDevice.cpp-
plugins/GSdx/GSDevice.cpp-	if (hit != m_pool.end()) {
plugins/GSdx/GSDevice.cpp-		GSTexture* t = *hit;
plugins/GSdx/GSDevice.cpp-		m_pool.erase(hit);
plugins/GSdx/GSDevice.cpp-		return t;
plugins/GSdx/GSDevice.cpp-	}
```
```
plugins/GSdx/GSTextureCache.cpp-	// Seach first in the same type cache
plugins/GSdx/GSTextureCache.cpp-	auto is_bp = [&bp](const Target* t) { return t->m_TEX0.TBP0 == bp; };
plugins/GSdx/GSTextureCache.cpp:	const auto hit = std::find_if(m_dst[type].begin(), m_dst[type].end(), is_bp);
plugins/GSdx/GSTextureCache.cpp-
plugins/GSdx/GSTextureCache.cpp-	if (hit != m_dst[type].end()) {
plugins/GSdx/GSTextureCache.cpp-		dst = *hit;
```

Comment (ssakash) [inline plugins/GSdx/GSDialog.cpp]: > Here some examples (because I played with the texture cache, I don't know if VS2013 support it).

I think lambda expressions were supported even at the VS2010 versions, so I _think_ it should compile.

File: plugins/GSdx/GSDialog.cpp [hunk 3]
```diff
@@ -166,10 +166,18 @@ void GSDialog::SetTextAsInt(UINT id, int i)
 
 void GSDialog::ComboBoxInit(UINT id, const vector<GSSetting>& settings, int32_t selectionValue, int32_t maxValue)
 {
+	if (settings.empty()) return;
+
 	HWND hWnd = GetDlgItem(m_hWnd, id);
 
 	SendMessage(hWnd, CB_RESETCONTENT, 0, 0);
 
+	auto is_present = [&selectionValue](const GSSetting x) { return (selectionValue == x.value); };
```
Comment (turtleli) [inline plugins/GSdx/GSDialog.cpp]: GSSetting should be passed by reference.
The parenthesis around `selectionValue == x.value` is unnecessary.
(`selectionValue` could be passed by value `[=]`.)

File: plugins/GSdx/GSDialog.cpp [hunk 4]
```diff
@@ -166,10 +166,18 @@ void GSDialog::SetTextAsInt(UINT id, int i)
...
+	bool currentsel_found = std::find_if(settings.begin(), settings.end(), is_present) != settings.end();
```
Comment (turtleli) [inline plugins/GSdx/GSDialog.cpp]: (The bool isn't necessary if the condition is just used directly in the if statement condition.)

**Comment by ssakash** [comment]
> (The bool isn't necessary if the condition is just used directly in the if statement condition.)

I think the bool variable's name makes the code self explanatory, I would prefer to have it.

```
const bool currentsel_found = std::find_if(settings.begin(), settings.end(), is_present) != settings.end();
if (!currentsel_found)
    selectionValue = settings.front().value;
```

**VS**

```
if(std::find_if(settings.begin(), settings.end(), is_present) == settings.end())
     selectionValue = settings.front().value;
```

**Comment by gregory38** [comment]
Well if you read the line you have `...find_if .... is_present... == ... end...`, it seems enough to understand that you check an entry isn't present in the container.

File: plugins/GSdx/GSDialog.cpp (hunk 3 above)
Comment (gregory38) [inline plugins/GSdx/GSDialog.cpp]: Hum, potentially a reference is easier for the compiler to optimize/inline the function.

**Comment by ssakash** [comment]
> Well if you read the line you have ...find_if .... is_present... == ... end..., it seems enough to understand that you check an entry isn't present in the container.

To be honest, I still think it's confusing without the variable name but _I don't feel like arguing right now_.

**Comment by turtleli** [comment]
Actually, now that I think about it, `std::none_of` might be a better fit and should be more readable.

**Comment by ssakash** [comment]
>Hum, potentially a reference is easier for the compiler to optimize/inline the function.

Oops, didn't notice your comment earlier. So what's the conclusion - pass by _reference_ **(or)** _value_?

**Comment by gregory38** [comment]
Well we don't really care. Do by value. (A const ref is better than a ref)

File: plugins/GSdx/GSDialog.cpp [hunk 5]
```diff
@@ -166,10 +166,17 @@ void GSDialog::SetTextAsInt(UINT id, int i)
...
 void GSDialog::ComboBoxInit(UINT id, const vector<GSSetting>& settings, int32_t selectionValue, int32_t maxValue)
 {
+	if (settings.empty())
+		return;
+
 	HWND hWnd = GetDlgItem(m_hWnd, id);
 
 	SendMessage(hWnd, CB_RESETCONTENT, 0, 0);
 
+	const auto is_present = [=](const GSSetting& x) { return selectionValue == x.value; };
```
Comment (gregory38) [inline plugins/GSdx/GSDialog.cpp]: Oh I didn't understand that `[=]` was enough to catch all variables by value. It is much nicer this way.

**Comment by gregory38** [comment]
For me it is fine.
